### PR TITLE
Port to python3

### PR DESCRIPTION
**Pull request by bemsha**

### CODE REVIEW

File: nd_service_registry/funcs.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 __author__ = 'matt@nextdoor.com (Matt Wise)'
 
-import cPickle as pickle
+import pickle as pickle
```
Comment (mrterry) [inline nd_service_registry/funcs.py]: cPickle is much faster.  you probably want to add six as a dep and `from six.moves import cPickle`

File: nd_service_registry/funcs_tests.py [hunk 2]
```diff
@@ -47,8 +47,8 @@ def test_decode_returns_string_value_dict_on_malformed_json_string_input(
 
     def test_default_data_produces_expected_dict(self):
         default_data = funcs.default_data()
-        self.assertIn(u"pid", default_data.keys())
-        self.assertIn(u"created", default_data.keys())
+        self.assertIn(u"pid", list(default_data.keys()))
```
Comment (mrterry) [inline nd_service_registry/funcs_tests.py]: for dicts, `in` inspects the keys.  so you don't even need `.keys()` here

**Comment by mrterry** [comment]
adding a py3 target for travis, would also be good

File: nd_service_registry/funcs_tests.py (hunk 2 above)
Comment (bemsha) [inline nd_service_registry/funcs_tests.py]: Done.

File: nd_service_registry/funcs.py (hunk 1 above)
Comment (bemsha) [inline nd_service_registry/funcs.py]: Done.

**Comment by mrterry** [comment]
`from __future__ import print_function` would be nice to have

File: nd_service_registry/bin/ndsr/ndsr.py [hunk 3]
```diff
@@ -19,8 +19,9 @@
 
 __author__ = 'me@ryangeyer.com (Ryan J. Geyer)'
```
Comment (mrterry) [inline nd_service_registry/bin/ndsr/ndsr.py]: i think this is making pep8 unhappy.

Comment (bemsha) [inline nd_service_registry/bin/ndsr/ndsr.py]: Why is print_function needed? Is there a difference that's important to us between python2 treating print's payload as a formatted thing and python3 treating it as an argument list?

Comment (mrterry) [inline nd_service_registry/bin/ndsr/ndsr.py]: `print_function` forces py3's syntax & semantics.  it keeps you from doing it the "wrong" way

**Comment by diranged** [comment]
@bemsha, not sure why your last few commits are Merge commits? Are you working on a fork -- or are you working in the main repo? Either is fine. Just curious.

File: nd_service_registry/bin/ndsr/ndsr.py (hunk 3 above)
Comment (bemsha) [inline nd_service_registry/bin/ndsr/ndsr.py]: Ok. I'll add that.

Comment (bemsha) [inline nd_service_registry/bin/ndsr/ndsr.py]: I disabled pep8's E402 error. The  __author__ setting will be worthwhile when we do more with OWNERS files.

**Comment by bemsha** [comment]
@diranged I don't have a good feeling for the workflow conventions people use on github. I made a branch from master on github and I've been pushing commit amendments to that branch.

**Comment by bemsha** [comment]
@diranged I don't know all the conventions for submissions to github. I followed Wesley's cookbook on Confluence. I made a branch from master on github and I've been pushing commit amendments to that branch.

File: nd_service_registry/funcs_tests.py [hunk 4]
```diff
@@ -47,8 +63,8 @@ def test_decode_returns_string_value_dict_on_malformed_json_string_input(
 
     def test_default_data_produces_expected_dict(self):
         default_data = funcs.default_data()
-        self.assertIn(u"pid", default_data.keys())
-        self.assertIn(u"created", default_data.keys())
+        self.assert_(u"pid" in default_data)
```
Comment (skrawcz) [inline nd_service_registry/funcs_tests.py]: change to assertTrue, assert_ is a deprecated alias apparently... (https://docs.python.org/3/library/unittest.html)

File: nd_service_registry/funcs_tests.py [hunk 5]
```diff
@@ -47,8 +63,8 @@ def test_decode_returns_string_value_dict_on_malformed_json_string_input(
...
+        self.assert_(u"created" in default_data)
```
Comment (skrawcz) [inline nd_service_registry/funcs_tests.py]: change to assertTrue, assert_ is a deprecated alias apparently... (https://docs.python.org/3/library/unittest.html)

**Comment by bemsha** [comment]
I've reset this change and force pushed it, to remove the unintended merges.

**Comment by bemsha** [comment]
@diranged I've asked you to review this as the original author, @mrterry to review it for python3, and Stefan to review it for usability. I've addressed all the comments. How do we move to closure?

**Comment by mrterry** [comment]
if you want real py3 compat, you need to always run the tests against py3.  please update `.travis.yml` to do so.  I think you just need to add a `- python 3.3` line in addition to the existing `- python 2.7` line.  once that happens, i'm happy

**Comment by bemsha** [comment]
I've added '- "3.3"' to .travis.yml.

**Comment by diranged** [comment]
@bemsha, Love that you're doing this ... but it looks like the tests still don't pass in Python 3.3. You can probably ignore the Kazoo 2.0 failures right now, as we know thats an ongoing issue (there is an issue for this somewhere) -- but we need to make the tests pass on Python 3.3 before merging. Also when you're finally done, squash the commits all into one. Thanks!

**Comment by bemsha** [comment]
Most of the failures are due to an error on my part. There are one or two
related to the kazoo API. I'll make this better.

      Chuck Karish             chuck@nextdoor.com               650 283 7864

**Comment by mikhail** [comment]
Solved din #67

**Pull request closed without merging**